test/tools/d_do_test.d

```diff
@@ -962,11 +965,76 @@ int runBashTest(string input_dir, string test_name)
     version(Windows)
     {
         auto process = spawnShell(format("bash %s %s %s",
-            buildPath(scriptDir, "tools", "sh_do_test.sh"), input_dir, test_name));
+            buildPath(dmdTestDir, "tools", "sh_do_test.sh"), input_dir, test_name));
     }
     else
     {
-        auto process = spawnProcess([scriptDir.buildPath("tools", "sh_do_test.sh"), input_dir, test_name]);
+        auto process = spawnProcess([dmdTestDir.buildPath("tools", "sh_do_test.sh"), input_dir, test_name]);
     }
     return process.wait();
 }
+
+int runDShellTest(string input_dir, string test_name, EnvData* envData,
+    string output_dir, string output_file)
+{
+    const testScriptPath = buildPath(dmdTestDir, input_dir, test_name ~ ".d");
+    const testScriptSource = readText(testScriptPath);
+    const dshellPath = buildPath(dmdTestDir, "tools", "dshell.d");
+    const dshellSource = readText(dshellPath);
+
+    writefln(" ... %s/%s.d", input_dir, test_name);
+    if (exists(output_file))
+        remove(output_file);
+
+    const testScriptExe = buildPath(output_dir, test_name ~ envData.exe);
+    auto output = appender!(char[])();
+    {
+        const compile = [envData.dmd, "-of" ~ testScriptExe, testScriptPath, dshellPath];
+        output.put("Compiling test...\n");
+        output.put(escapeShellCommand(compile));
+        output.put("\n");
+        const compileResult = std.process.execute(compile);
+        output.put(compileResult.output);
+        if (!compileResult.output.endsWith('\n'))
+            output.put("\n");
+        if (compileResult.status != 0)
+        {
+            dshellTestFailed(output.data, input_dir, test_name);
+            return compileResult.status;
+        }
+    }
+    {
+        const runTest = [testScriptExe, input_dir, test_name];
+        output.put("Running test...\n");
+        output.put(escapeShellCommand(runTest));
+        output.put("\n");
+        const runResult = std.process.execute(runTest);
+        output.put(runResult.output);
+        if (!runResult.output.endsWith('\n'))
+            output.put("\n");
+        if (runResult.status != 0)
+        {
+            dshellTestFailed(output.data, input_dir, test_name);
+            return runResult.status;
+        }
+    }
+    mkdirFor(output_file);
+    auto file = File(output_file, "w");
+    file.write(output.data);
+    return 0; // success
+}
+auto dshellTestFailed(const(char)[] output, string input_dir, string test_name)
+{
+    writeln("==============================");
+    writefln("Test '%s/%s.d' failed. The logged output:", input_dir, test_name);
+    write(output);
+    if (!output.endsWith("\n"))
+          writeln();
+    writeln("==============================");
+}
+void mkdirFor(string filename)
```

Comment:
DStyle: short description of the function + one line of whitespace between functions